### PR TITLE
Expose Adjustment Rule BaseUtcOffsetDelta

### LINKED ISSUE #50256: Add BaseUtcOffset to TimeZoneInfo.AdjustmentRule.
*This proposal edited to address the feedback we got in the design review*

## Background and Motivation

TimeZoneInfo.Adjustment provides information about a time zone adjustment, such as the transition to and from daylight saving time. every adjustment rule has a start and end dates to describe at what date and time the rule can apply. Although the rule provides the information that help convert from the local time to UTC and vice versa, but it lacks an essential information which is what is the base UTC offset for the time zone during the rule period. TimeZoneInfo originally assumed that the base UTC offset for the zone is constant and never change which is incorrect. There are zones changed the UTC offset at some point. 

The proposal here is have the AdjustmentRule class provide a property telling what the base UTC offset delta of the zone during the rule start and end date is. 

## Proposed API

```C#
    public sealed partial class TimeZoneInfo
    {
        public sealed class AdjustmentRule : IEquatable<AdjustmentRule?>, ISerializable, IDeserializationCallback
        {
             public TimeSpan BaseUtcOffsetDelta { get; }

             public static TimeZoneInfo.AdjustmentRule CreateAdjustmentRule(
                              DateTime dateStart, 
                              DateTime dateEnd, 
                              TimeSpan daylightDelta, 
                              TimeZoneInfo.TransitionTime daylightTransitionStart, 
                              TimeZoneInfo.TransitionTime daylightTransitionEnd, 
                              TimeSpan baseUtcOffsetDelta); // this is the only added parameter compared to the other overload.
        }
    }
```

## Usage Examples

```C#
    // On 27 March 2011, Russia moved to year-round daylight-saving time. Instead of switching between UTC+09:00 in winter and UTC+10:00 in summer, 
    // Yakutsk Time became fixed at UTC+10:00 until 2014, when it was reset back to UTC+09:00 year-round.
    TimeZoneInfo tzi = TimeZoneInfo.FindSystemTimeZoneById("Asia/Yakutsk");
    foreach (TimeZoneInfo.AdjustmentRule rule in tzi.GetAdjustmentRules())
    {
        Console.WriteLine($"Base Utc offset during the time {rule.DateStart} and {rule.DateEnd} is {tzi.BaseUtcOffset + rule.BaseUtcOffsetDelta}");
    }
```

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes #50256

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @mattjohnsonpint @eerhardt @safern

**Comment by tarekgh** [comment]
CC @gewarren for the doc part.

File: src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs [hunk 1]
```diff
@@ -2314,7 +2314,7 @@ public static IEnumerable<object[]> SystemTimeZonesTestData()
                 for (int i = -14; i <= 12; i++)
                 {
                     TimeZoneInfo tz = null;
- 
+
                     try
                     {
                         string id = $"Etc/GMT{i:+0;-0}";
```
Comment (stephentoub) [inline src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs:2320]: This formatting is going to be subject to the current culture, which could impact the negative sign used.  Is that going to be an issue for someone running the tests in, say, Sweden?

**Review by stephentoub — APPROVED** [review]

**Review by eerhardt — APPROVED** [review]
:shipit:

File: src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs (hunk 1 above)
Comment (tarekgh) [reply src/libraries/System.Runtime/tests/System/TimeZoneInfoTests.cs:2320]: It is possible to fail. As this is not related to this PR, we can address that in another PR.

---
In reply to: [611643769](https://github.com/dotnet/runtime/pull/51055#discussion_r611643769) [](ancestors = 611643769)

File: src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.AdjustmentRule.cs [hunk 2]
```diff
@@ -76,7 +79,17 @@ public sealed class AdjustmentRule : IEquatable<AdjustmentRule?>, ISerializable,
...
+            /// <param name="dateEnd">The last date that the adjustment rule is in force. If the value of the dateEnd parameter is DateTime.MaxValue.Date, the adjustment rule has no end date.</param>
+            /// <param name="daylightDelta">The time change that results from the adjustment. This value is added to the time zone's BaseUtcOffset+BaseUtcOffsetDelta properties to obtain the correct daylight offset from Coordinated Universal Time (UTC). This value can range from -14 to 14.</param>
+            /// <param name="daylightTransitionStart">An object that defines the start of daylight saving time.</param>
+            /// <param name="daylightTransitionEnd">An object that defines the end of daylight saving time.</param>
+            /// <param name="baseUtcOffsetDelta">The time difference with the particular time zone base Utc offset during this adjustment rule period.</param>
+            /// <returns>An object that represents the new adjustment rule.</returns>
```
Comment (gewarren) [inline src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.AdjustmentRule.cs:91]:
```suggestion
            /// <returns>The new adjustment rule.</returns>
```

File: src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.AdjustmentRule.cs [hunk 3]
```diff
@@ -31,7 +31,10 @@ public sealed class AdjustmentRule : IEquatable<AdjustmentRule?>, ISerializable,
 
             public TransitionTime DaylightTransitionEnd => _daylightTransitionEnd;
 
-            internal TimeSpan BaseUtcOffsetDelta => _baseUtcOffsetDelta;
+            /// <summary>
+            /// Gets the time difference with the particular time zone base Utc offset during the adjustment rule period.
```
Comment (gewarren) [inline src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.AdjustmentRule.cs:35]:
```suggestion
            /// Gets the time difference with the base UTC offset for the time zone during the adjustment-rule period.
```

File: src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.AdjustmentRule.cs [hunk 4]
```diff
@@ -76,7 +79,17 @@ public sealed class AdjustmentRule : IEquatable<AdjustmentRule?>, ISerializable,
                 _noDaylightTransitions = noDaylightTransitions;
             }
 
-            internal static AdjustmentRule CreateAdjustmentRule(
+            /// <summary>
+            /// Creates a new adjustment rule for a particular time zone.
+            /// </summary>
+            /// <param name="dateStart">The effective date of the adjustment rule. If the value of the dateStart parameter is DateTime.MinValue.Date, this is the first adjustment rule in effect for a time zone.</param>
```
Comment (gewarren) [inline src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.AdjustmentRule.cs:85]:
```suggestion
            /// <param name="dateStart">The effective date of the adjustment rule. If the value is <c>DateTime.MinValue.Date</c>, this is the first adjustment rule in effect for a time zone.</param>
```

File: src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.AdjustmentRule.cs [hunk 5]
```diff
@@ -76,7 +79,17 @@ public sealed class AdjustmentRule : IEquatable<AdjustmentRule?>, ISerializable,
...
+            /// <param name="dateEnd">The last date that the adjustment rule is in force. If the value of the dateEnd parameter is DateTime.MaxValue.Date, the adjustment rule has no end date.</param>
```
Comment (gewarren) [inline src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.AdjustmentRule.cs:86]:
```suggestion
            /// <param name="dateEnd">The last date that the adjustment rule is in force. If the value is <c>DateTime.MaxValue.Date</c>, the adjustment rule has no end date.</param>
```

File: src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.AdjustmentRule.cs [hunk 6]
```diff
@@ -76,7 +79,17 @@ public sealed class AdjustmentRule : IEquatable<AdjustmentRule?>, ISerializable,
...
+            /// <param name="daylightDelta">The time change that results from the adjustment. This value is added to the time zone's BaseUtcOffset+BaseUtcOffsetDelta properties to obtain the correct daylight offset from Coordinated Universal Time (UTC). This value can range from -14 to 14.</param>
```
Comment (gewarren) [inline src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.AdjustmentRule.cs:87]:
```suggestion
            /// <param name="daylightDelta">The time change that results from the adjustment. This value is added to the time zone's <see cref="P:System.TimeZoneInfo.BaseUtcOffset" /> and <see cref="P:System.TimeZoneInfo.BaseUtcOffsetDelta" /> properties to obtain the correct daylight offset from Coordinated Universal Time (UTC). This value can range from -14 to 14.</param>
```

File: src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.AdjustmentRule.cs [hunk 7]
```diff
@@ -76,7 +79,17 @@ public sealed class AdjustmentRule : IEquatable<AdjustmentRule?>, ISerializable,
...
+            /// <param name="daylightTransitionStart">An object that defines the start of daylight saving time.</param>
```
Comment (gewarren) [inline src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.AdjustmentRule.cs:88]:
```suggestion
            /// <param name="daylightTransitionStart">The start of daylight saving time.</param>
```

File: src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.AdjustmentRule.cs [hunk 8]
```diff
@@ -76,7 +79,17 @@ public sealed class AdjustmentRule : IEquatable<AdjustmentRule?>, ISerializable,
...
+            /// <param name="daylightTransitionEnd">An object that defines the end of daylight saving time.</param>
```
Comment (gewarren) [inline src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.AdjustmentRule.cs:89]:
```suggestion
            /// <param name="daylightTransitionEnd">The end of daylight saving time.</param>
```

File: src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.AdjustmentRule.cs [hunk 9]
```diff
@@ -76,7 +79,17 @@ public sealed class AdjustmentRule : IEquatable<AdjustmentRule?>, ISerializable,
...
+            /// <param name="baseUtcOffsetDelta">The time difference with the particular time zone base Utc offset during this adjustment rule period.</param>
```
Comment (gewarren) [inline src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.AdjustmentRule.cs:90]:
```suggestion
            /// <param name="baseUtcOffsetDelta">The time difference with the base UTC offset for the time zone during the adjustment-rule period.</param>
```

**Review by gewarren — COMMENTED** [review]
I left some /// suggestions only.

**Comment by mattjohnsonpint** [comment]
LGTM

**Pull request merged**